text/0000-nested-publish.md

```diff
@@ -0,0 +1,352 @@
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+## Cargo: Manifest
+
+Two new possible values are added to the manifest.
+
+*   Packages may be specified as eligible for nested publishing using the `package.publish.nested` (or `workspace.package.publish.nested`) field, which takes the values `true` or `false` and defaults to `false`. `true` permits the package to be a nested package as defined in this RFC, affecting `cargo publish` and builds as discussed below, and `false` prohibits it as is the status quo.
+
+    ```toml
+    [package]
+    # This package may be nested but may not be `cargo publish`ed by itself
+    publish.nested = true
+    ```
+
+
+    To allow both regular publishing and nested publishing, the currently-allowed values for `publish` (`publish = true` or `publish = [registries...]`) may be specified under the key `package.publish.registries`. Examples:.
+
+    ```toml
+    [package]
+    publish.registries = ["crates.io"]
+    publish.nested = true
+    ```
+
+    ```toml
+    [package]
+    publish.registries = true
+    publish.nested = true
+    ```
+
+    If `package.publish` is a table, then `package.publish.registries` defaults to `false`, regardless of the value or presence of `package.publish.nested`.
+
+    Note: This dual-publishing-mode functionality is permitted mainly to keep the functionality composable/orthogonal. We hope that in most cases, packages are either published nested exactly once, or to a registry alone, to avoid duplicating code in the registry and compiling it redundantly.
+
+* The `dependencies.*.publish` field is newly defined, with the only currently allowed value being `"nested"`, to declare that that dependency is a nested dependency.
+    * It is an error if a nested dependency does not have a `path` field, or if it has a `version`, `git`, or any other package source field, unless future work defines a meaning for that combination.
+    * Workspace inheritance is not permitted; the presence of `workspace.dependencies.*.publish` is an error.
+
+When a nested dependency is present (making its referent be a nested package), the following additional requirements apply:
+
+* The nested package must have `package.publish.nested = true`.
+* If the nested package specifies `package.license`, its value must be identical to the parent package's.
+
+  This check is intended only to prevent accidents (such as vendoring a third-party package without considering the implications of redistributing it). It is always valid to omit `package.license` from the nested package, thus making no machine-readable claims about its licensing.
+
+It is an error for a nested package to have the same package name as the parent package or any other nested package with the same parent package. This is validated by all Cargo operations that would generate or read a lockfile.
```

Comment:
What is the rationale for this?  I'm assuming this is for when we move all of the packages into the same location but it'd be good to be explicit about it.

Reply:
> I'm assuming this is for when we move all of the packages into the same location

Yes.

I also expect there will be various other cases where we want to name the packages uniquely, including ones where it's not ideal but may be done temporarily for the first round of implementation, and by imposing this restriction, it's always sufficient to use some form of a (parent package name, nested package name) tuple.

As precedent, Cargo currently prohibits `path` dependencies with the same package name. But we do need to allow nested package names to be the same when they're in different parent packages, since otherwise packages could conflict with each other due to their implementation details. (I added a paragraph on that.)